### PR TITLE
refactor(AllRaids): Add new AllRaids page

### PR DESCRIPTION
**Pull request by datner**
theres a lot going on, so take your time

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,7 +18,8 @@
     "firebase": "^8.2.4",
     "react": "^17.0.0",
     "react-dom": "^17.0.0",
-    "react-router-dom": "^5.2.0"
+    "react-router-dom": "^5.2.0",
+    "reactfire": "^3.0.0-rc.0"
```
Comment (datner) [inline package.json:22]: note to self: blam this

File: src/styles/globalStyles.ts [hunk 2]
```diff
@@ -37,15 +38,15 @@ const globalStyles = css`
     --h4: calc(var(--h5) / var(--text-scale-ratio)); // 25
     --h5: calc(1em / var(--text-scale-ratio)); // 20
     --small-text: calc(1em * var(--text-scale-ratio)); // 12.8
-    @media screen and (max-width: 1024px) {
+    @media screen and (max-width: 1023px) {
       --text-scale-ratio: 0.85;
     }
-    @media screen and (max-width: 768px) {
+    @media screen and (max-width: 767px) {
```
Comment (datner) [inline src/styles/globalStyles.ts:44]: when using max-width, -1px is required

File: src/styles/globalStyles.ts [hunk 3]
```diff
@@ -37,15 +38,15 @@ const globalStyles = css`
...
       --text-scale-ratio: 0.9;
     }
 
     /* Spacing (fibonacci) */
-    --space-unit: 1em;
+    --space-unit: 1rem;
```
Comment (JacobMGEvans) [inline src/styles/globalStyles.ts:49]: Why was this changed, last time I checked the guild had chosen `em` after ad nauseam discussion.  @nobrayner @merodiro

File: src/utils/useCollection/index.ts [hunk 4]
```diff
@@ -1,52 +1,48 @@
 import { useState, useEffect } from 'react'
-import firestore from '#utils/useFirestore'
+import firestore from '#utils/firestore'
```
Comment (JacobMGEvans) [inline src/utils/useCollection/index.ts:2]: Why was the naming convention changed?

File: src/routes/raids/ViewRaid.tsx [hunk 5]
```diff
@@ -19,7 +19,7 @@ const userStatSortNames = Object.keys(userStatSorts)
 
 const ViewRaid = () => {
   const { raidId } = useParams<{ raidId: string }>()
-  const documentData = useDocument<ViewRaidData>('raid-stats', raidId)
+  const documentData = useDocument<Raid>('raid-stats', raidId)
```
Comment (JacobMGEvans) [inline src/routes/raids/ViewRaid.tsx:22]: Why was this changed?

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (JacobMGEvans) [reply src/styles/globalStyles.ts:44]: for what?

File: src/utils/useDocument/index.ts [hunk 6]
```diff
@@ -1,5 +1,5 @@
 import { useState, useEffect } from 'react'
-import firestore from '#utils/useFirestore'
+import firestore from '#utils/firestore'
```
Comment (JacobMGEvans) [inline src/utils/useDocument/index.ts:2]: Same as above

File: types/viewRaidData.d.ts [hunk 7]
```diff
@@ -7,7 +7,7 @@ interface UserStats {
   commits: number
 }
 
-interface ViewRaidData {
+interface Raid {
```
Comment (JacobMGEvans) [inline types/viewRaidData.d.ts:10]: Why the naming convention change?

File: src/routes/raids/ViewRaid.tsx (hunk 5 above)
Comment (nobrayner) [reply src/routes/raids/ViewRaid.tsx:22]: I'll admint `ViewRaidData` wasn't the best of names... I think `RaidData` would make more sense than `Raid` in this case, though

File: src/styles/globalStyles.ts (hunk 3 above)
Comment (nobrayner) [reply src/styles/globalStyles.ts:49]: I believe this comes down to a difference in what these values are intended to be used for, and how people are perceiving they should be used for.

For example, an `em` is probably not your best choice if you are going to use it for general "pad this, margin that" - which is what seems people are perceiving they can use this for.

This whole "how is CSS handled" discussion I think needs to be "run through the mob" again, methinks. Specifically, we need to get an agreement of what/how the system will work, document it, and then from there it can be more easily enforced - or changes discussed if there is a valid case for changing it.

File: src/styles/globalStyles.ts [hunk 8]
```diff
@@ -56,16 +57,26 @@ const globalStyles = css`
     @media screen and (max-width: 1024px) {
       --space-unit: 0.8em;
     }
+
+    /* Shadows */
+    --elevation-1: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
```
Comment (nobrayner) [inline src/styles/globalStyles.ts:62]: I'd be curious of others opinions on this: it took me a few seconds to understand what an "elevation" css property was for - which would be a shadow that gives the appearance of a certain level of elevation.

Would it make more sense to just call these `shadow` for future ease of use? Or is elevation already an established convention, and I am just behind :sweat_smile:

File: src/utils/firestore/index.ts [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import firebase from 'firebase/app'
 import 'firebase/firestore'
 
-const firebaseConfig = {
+export const firebaseConfig = {
```
Comment (nobrayner) [inline src/utils/firestore/index.ts:4]: This shouldn't need to be exported... It's intentionally only exposing `firestore` - you shouldn't be using the config anywhere else.

File: src/utils/useCollection/index.ts [hunk 10]
```diff
@@ -1,52 +1,48 @@
...
+              id: doc.id,
+              ...doc.data(),
+            } as WithID<TDocument>),
+        )
+        setCollectionState({ data: documents, loading: false, error: null })
+      } catch (error) {
+        setCollectionState({ data: [], loading: false, error })
+      }
+    }
+    getCollection()
```
Comment (nobrayner) [inline src/utils/useCollection/index.ts:42]: Why change what wasn't broken? :grey_question: 

Also, you have now lost a case of distinguishing a "Not found" and an error - probably didn't _quite_ make sense for the collection hook, but still...

File: src/styles/globalStyles.ts (hunk 8 above)
Comment (codyarose) [reply src/styles/globalStyles.ts:62]: @nobrayner It's a convention I've seen elsewhere. Basically a higher elevation number should mean that the element appears to 'float' at a higher elevation off of the page/screen - I think it's originally a material design convention. Obviously it should be renamed if different elevations aren't part of the design.

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (codyarose) [reply src/styles/globalStyles.ts:44]: Yeah, it should be the other way around. 
max-width for tablet size should be 1024 (or 64r/em), min-width would be +1px

File: src/styles/globalStyles.ts (hunk 3 above)
Comment (codyarose) [reply src/styles/globalStyles.ts:49]: As I mentioned in discord, separate flexible and fixed spacing scales would be better than forcing it all to be rem-based. `--space-1` being `em` and `--space-fixed-1` being `rem`

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (datner) [reply src/styles/globalStyles.ts:44]: min width = "as long as the width is this number or higher, these apply" or, px > min-width
max width = "as long as the width is lower than or equal to this number, these apply, or, px =< max-height

therefore, we need to remove an extra-pixel. But don't take my word for it, use chrome devtools and play around with the tablet sizes

File: src/utils/useCollection/index.ts (hunk 4 above)
Comment (datner) [reply src/utils/useCollection/index.ts:2]: it's not a hook, on the convention did not change, it was fixed

File: src/routes/raids/ViewRaid.tsx (hunk 5 above)
Comment (datner) [reply src/routes/raids/ViewRaid.tsx:22]: Suffixing `Data` to stuff is superfluous, most things are data, it's the other way around that we want to tag it.
for example, when making a database model of a raid, would you call the entity `RaidData`? or `Raid`?

File: src/styles/globalStyles.ts (hunk 8 above)
Comment (datner) [reply src/styles/globalStyles.ts:62]: elevation is borrowed from [material design](https://material.io/design/environment/elevation.html), it's the most popular design methodology so it's safe to assume that most devs would recognise it. But it also gives space to other shadow-themed variables without confusion

File: src/utils/firestore/index.ts (hunk 9 above)
Comment (datner) [reply src/utils/firestore/index.ts:4]: artefacts from when I tried to work with `reactfire`, the config shouldn't be committed in the first place if you want to actually discuss this

File: src/utils/useCollection/index.ts (hunk 10 above)
Comment (datner) [reply src/utils/useCollection/index.ts:42]: always leave things cleaner than you found them, it was very verbose and made a distinction (not-found) that doesn't actually reflect a real state. when accessing firestore with a bogus key, you just receive an empty result, and using a good key in an empty collection gives the same result. either way you can just check the data.length instead of mimicking a state machine

meaning that possible states, when given names, is loading, success, success_empty (or not found, if you will), and error.

File: src/styles/globalStyles.ts (hunk 3 above)
Comment (datner) [reply src/styles/globalStyles.ts:49]: anywho, i need something so space with, and if I change this back to relative, the I will need to use static numbers in places where this was implemented, otherwise the design breaks, so suggestions are welcome

File: src/utils/useCollection/index.ts (hunk 10 above)
Comment (datner) [reply src/utils/useCollection/index.ts:42]: also, I did not lose it, I can assign `data: null` on error and make the type of data nullable, but that drags a whole load of aggressive type guarding, I've tried it, it looks bad and is annoying to work with. You just need to handle loading -> error -> data which is convention in [apollo client](https://www.apollographql.com/docs/react/get-started/#request-data) [react-query](https://react-query.tanstack.com/guides/queries) [use-http](https://use-http.com/#/?id=basic-usage-auto-managed-state)  [swr](https://swr.vercel.app/) and many many more. So it's not contrived to work assuming this workflow

File: src/utils/firestore/index.ts (hunk 9 above)
Comment (JacobMGEvans) [reply src/utils/firestore/index.ts:4]: In Firebase documentation the frontend config can be exposed, if you actually want to discuss this. https://firebase.google.com/docs/web/setup

File: types/viewRaidData.d.ts (hunk 7 above)
Comment (JacobMGEvans) [reply types/viewRaidData.d.ts:10]: The convention is what was already there, regardless if you found it superfluous or not. 

As for changing things that is always up for discussion.

File: src/routes/raids/ViewRaid.tsx (hunk 5 above)
Comment (JacobMGEvans) [reply src/routes/raids/ViewRaid.tsx:22]: I'd rather the suffix be there, maybe `RaidStats` instead. There will be more things in the future related to Raid and that will get convoluted fast.

File: types/viewRaidData.d.ts (hunk 7 above)
Comment (JacobMGEvans) [reply types/viewRaidData.d.ts:10]: As above suggested `RaidStats`

File: src/styles/globalStyles.ts (hunk 8 above)
Comment (JacobMGEvans) [reply src/styles/globalStyles.ts:62]: I've seen this convention as well, specifically from Material UI

File: src/utils/useCollection/index.ts (hunk 10 above)
Comment (JacobMGEvans) [reply src/utils/useCollection/index.ts:42]: I rather mimic a finite state machine. This only "looks cleaner" in my opinion and that's not a big priority. The event state machine esque is more declarative and explicit of a pattern, so will look busier, i.e. XState configs look busy and messy, but are extremely explicit and declarative, it's kind of the point.

File: src/styles/globalStyles.ts (hunk 3 above)
Comment (JacobMGEvans) [reply src/styles/globalStyles.ts:49]: > it might sound dumb, but if I'm writing `--space-fixed-1` ten times more than `--space-1` then i'd rather it'll be the other way around: `--space-1` and `--space-relative-1`

This just sounds like personal preference, about something already decided.

Comment (datner) [reply src/styles/globalStyles.ts:49]: that is entirely correct

File: src/routes/raids/ViewRaid.tsx (hunk 5 above)
Comment (datner) [reply src/routes/raids/ViewRaid.tsx:22]: `RaidStats` sound good, doing that

File: src/utils/firestore/index.ts (hunk 9 above)
Comment (datner) [reply src/utils/firestore/index.ts:4]: >Note: The Firebase config object contains unique, but non-secret identifiers for your Firebase project.
Visit Understand Firebase Projects to learn more about this config object.

I stand corrected!

File: src/utils/useCollection/index.ts (hunk 10 above)
Comment (datner) [reply src/utils/useCollection/index.ts:42]: we don't get a more explicit or descriptive state either way, the state returned from the firebase is either success with 0-n elements, or error, I don't see the need to go further than a regular `async` function for such a small pool of states (pending, success, error)

File: src/utils/useCollection/index.ts (hunk 4 above)
Comment (JacobMGEvans) [reply src/utils/useCollection/index.ts:2]: Lets make it one then 
https://usehooks.com/useFirestoreQuery/

File: src/utils/useDocument/index.ts (hunk 6 above)
Comment (JacobMGEvans) [reply src/utils/useDocument/index.ts:2]: Same as above. 
https://usehooks.com/useFirestoreQuery/

File: src/utils/useCollection/index.ts (hunk 4 above)
Comment (JacobMGEvans) [reply src/utils/useCollection/index.ts:2]: This wont block anything from being merged, but can be a TODO for sure

File: src/utils/useCollection/index.ts (hunk 10 above)
Comment (JacobMGEvans) [reply src/utils/useCollection/index.ts:42]: The tooling we generally use React-Query and XState would disagree with that, so I am all for finding a middle ground here but this came down to your refactoring based on your preference. 

Cleaner or clever refactors aren't always better. I leave this up to @nobrayner it was his original code. 

I would like to see tests for Refactors though.

File: src/styles/globalStyles.ts (hunk 8 above)
Comment (nobrayner) [reply src/styles/globalStyles.ts:62]: Cool - I learned something new :+1:

File: src/utils/firestore/index.ts (hunk 9 above)
Comment (nobrayner) [reply src/utils/firestore/index.ts:4]: Just remove the export, and this would be A-OK for me

```suggestion
const firebaseConfig = {
```

File: src/utils/useCollection/index.ts (hunk 10 above)
Comment (nobrayner) [reply src/utils/useCollection/index.ts:42]: The biggest issue I have here - and it doesn't relate directly to this code anymore - is that there is no longer a distinction between a "not found/no items" and error state. I'll make a new review on that to more focus the discussion

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (nobrayner) [reply src/styles/globalStyles.ts:44]: I'd agree with Cody here - a difference of one pixel really won't make a difference... So please use a round number :grimacing:

File: src/styles/globalStyles.ts (hunk 3 above)
Comment (nobrayner) [reply src/styles/globalStyles.ts:49]: Resolving this, as it can be further discussed in the CSS discussion we have opened.

https://github.com/OpenSourceRaidGuild/website/discussions/79

File: types/viewRaidData.d.ts [hunk 11]
```diff
@@ -7,7 +7,7 @@ interface UserStats {
   commits: number
 }
 
-interface ViewRaidData {
+interface RaidStats {
```
Comment (nobrayner) [inline types/viewRaidData.d.ts:10]: Should probably rename the file to `raidStats.d.ts` while we are at it :thinking:

**Review by nobrayner — CHANGES_REQUESTED** [review]

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (datner) [reply src/styles/globalStyles.ts:44]: what, this is not a matter of opinion lmao, it's the standard viewport of devices, it either works or doesn't, it didn't so I fixed it 😅 😆

Comment (datner) [reply src/styles/globalStyles.ts:44]: either that, to switch to min-width, like I do in my own components btw

File: src/utils/useCollection/index.ts (hunk 4 above)
Comment (datner) [reply src/utils/useCollection/index.ts:2]: this hook is essentially what is implemented in `useCollection` logic-wise, I found that there is no use for a generic access to firestore............ _for now_

File: .gitignore [hunk 12]
```diff
@@ -1,4 +1,6 @@
 .snowpack
+.vscode/
```
[the remark this answers is not in the document]
Comment (datner) [reply .gitignore:2]: I actually changed to vscode to be inline with how you guys work, and some of my extensions there started editing files in .vscode, I can can exclude the specific file that was edited by my instance instead of the entire folder, on it

File: src/routes/raids/index.tsx [hunk 13]
```diff
@@ -1,54 +1,258 @@
...
-                </Link>
-              </li>
-            ))}
-        </ul>
-      </>
+      <p>
+        {`Strange... I could've sworn the Manticore was here a second ago! Seems
+        there aren't any Raids right now. Try again later!`}
+      </p>
     )
```
[the remark this answers is not in the document]
Comment (datner) [reply src/routes/raids/index.tsx:48]: not-found does not return an error though, it returns an empty array from my tests

File: types/viewRaidData.d.ts (hunk 11 above)
Comment (datner) [reply types/viewRaidData.d.ts:10]: on it

File: src/routes/raids/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (datner) [reply src/routes/raids/index.tsx:48]: I can add a `!data.length` check if that is what you mean

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (codyarose) [reply src/styles/globalStyles.ts:44]: Well aware of what max and min width are, and tablet sizes. If I want to change styles for tablet size and down it would be `max-width: 1024px` because that is the viewport width of the tablet. If the max width is set to 1023 the styles won't be applied

File: src/routes/raids/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (nobrayner) [reply src/routes/raids/index.tsx:48]: > not-found does not return an error though

Correct. The way you have it currently though, is that when there is an error, you are showing what I had previously as the "data.length === 0" case. Which means the actual error is being suppressed.

If you could add `!data.length` checks, and also a short message about there being no data, that would be perfect; combined with rendering the error message in the error state.

I hope I explained better that time? :sweat_smile:

File: src/utils/useCollection/index.ts (hunk 4 above)
Comment (JacobMGEvans) [reply src/utils/useCollection/index.ts:2]: useCollection is specific, I would rather there is a `useFirebase` and is **composed** in more specific Custom Hooks like `useCollection`, `useCharacters`, `useWhateverFirebaseData`, or whatever. Likely the general Firebase logic was put into `useCollection` due to when it was written it being the only Firebase Query

Comment (nobrayner) [reply src/utils/useCollection/index.ts:2]: There are two query style hooks so far - `useDocument` and `useCollection` - useCollection can probably be turned into a `useFirestoreQuery` hook keeping all the same logic.

`useDocument` could then be re-written to utilize `useFirestoreQuery` under-the-hood, but I think it will still be a useful abstraction for just grabbing one document by it's id - though I can understand not wanting it at all :stuck_out_tongue:

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (datner) [reply src/styles/globalStyles.ts:44]: so this is the desired bahavior?
```css
.box {
 color: red;
}
@media (max-width: 1024px) {
  .box {
    color: blue;
  }
}
```
so this is the desired result?
- phone: blue
- tablets: blue
- tablets + 1px: red
- large tablets: red
- laptop: blue
- PC: blue

I'd think that the result would be:
- phone: blue
- tablets: red
- tablets + 1px: red
- large tablets: red
- laptop: blue
- PC: blue

File: src/routes/raids/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (datner) [reply src/routes/raids/index.tsx:48]: yeah, support an empty/not found case aswell

File: src/styles/globalStyles.ts (hunk 2 above)
Comment (nobrayner) [reply src/styles/globalStyles.ts:44]: This whole conversation is a bit pointless now, IMO.

They are just arbitrary numbers... Can we stop bikeshedding now? Thanks :sweat_smile: :100:

If you look at just Apple tablets, then there isn't even a clear consensus... https://yesviz.com/tablets.php So can we please just pick `1024` to appease my round (rem) number needs?
